File: app/src/main/java/com/example/findmeinlol/RiotAPIRepository.java
package com.example.findmeinlol;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import androidx.lifecycle.MutableLiveData;

import com.example.findmeinlol.model.Item;
import com.example.findmeinlol.model.data.InfoDto;
import com.example.findmeinlol.model.data.LeagueEntryDTO;
import com.example.findmeinlol.model.data.MatchDto;
import com.example.findmeinlol.model.data.ParticipantDto;
import com.example.findmeinlol.model.data.PerksDto;
import com.example.findmeinlol.model.data.SummonerDto;
import com.google.gson.Gson;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.concurrent.TimeUnit;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RiotAPIRepository {
    private static RiotAPIRepository mRetrofitRepository;

    static final String RIOT_URL = "https://kr.api.riotgames.com/lol/";
    static final String RIOT_MATCH_URL = "https://asia.api.riotgames.com/";
    static final String RIOT_PROFILE_IMAGE_URL =
            "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/profileicon/";
    static final String RIOT_CHAMPION_IMAGE_URL =
            "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/champion/";
    static final String RIOT_ITEM_IMAGE_URL =
            "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/item/";
    static final String RIOT_SPELL_IMAGE_URL =
            "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/spell/";
    static final String RIOT_SPELL_JSON_URL =
            "https://ddragon.leagueoflegends.com/cdn/13.5.1/data/en_US/";
    static final String RIOT_RUNES_JSON_URL =
            "https://ddragon.leagueoflegends.com/cdn/13.5.1/data/en_US/";
    static final String RIOT_RUNES_IMAGE_URL =
            "https://ddragon.leagueoflegends.com/cdn/img/";

    public MutableLiveData<SummonerDto> summonerDtoMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<ParticipantDto> participantDtoMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<Integer> integerMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<Item> championIconMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<Item> itemIconMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<Item> spellIconMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<Item> runeIconMutableLiveData = new MutableLiveData<>();
    private ArrayList<Integer> itemIds = new ArrayList<>();
    private ArrayList<Integer> spellIds = new ArrayList<>();
    private ArrayList<Integer> runeIds = new ArrayList<>();
    private HashMap<String, String> spellNames = new HashMap<>();
    private HashMap<Integer, String> runeNames = new HashMap<>();


    public static RiotAPIRepository getInstance() {
        if (mRetrofitRepository == null) {
            mRetrofitRepository = new RiotAPIRepository();
        }
        return mRetrofitRepository;
    }

    public RiotAPI mRiotApi = new Retrofit.Builder()
            .baseUrl(RIOT_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotMatchApi = new Retrofit.Builder()
            .baseUrl(RIOT_MATCH_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotProfileIconApi = new Retrofit.Builder()
            .baseUrl(RIOT_PROFILE_IMAGE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotChampionImgApi = new Retrofit.Builder()
            .baseUrl(RIOT_CHAMPION_IMAGE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotItemImgApi = new Retrofit.Builder()
            .baseUrl(RIOT_ITEM_IMAGE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotSpellImgApi = new Retrofit.Builder()
            .baseUrl(RIOT_SPELL_IMAGE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotSpellJSonApi = new Retrofit.Builder()
            .baseUrl(RIOT_SPELL_JSON_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotRunesJsonApi = new Retrofit.Builder()
            .baseUrl(RIOT_RUNES_JSON_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    public RiotAPI mRiotRuneImgApi = new Retrofit.Builder()
            .baseUrl(RIOT_RUNES_IMAGE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(RiotAPI.class);

    private interface ApiListener {
        void onUpdated();
    }

    private ApiListener apiListener = new ApiListener() {
        int cnt = 0;
        @Override
        public void onUpdated() {
            ParticipantDto participantDto = participantDtoMutableLiveData.getValue();

            getChampionImage(participantDto.getChampionName(), participantDto.getMatchId());
            for (int i = 0; i < 7 ; i++) {
                getItemImage(itemIds.get(i), participantDto.getMatchId(), i);
            }
            for (int i = 0; i < 2; i++) {
                getSpellImage(spellIds.get(i), participantDto.getMatchId(), i);
            }

            for (int i = 0; i < 2; i++) {
                getRuneImage(runeIds.get(i), participantDto.getMatchId(), i);
            }

            itemIds.clear();
            spellIds.clear();
            runeIds.clear();
            cnt++;

            if (cnt == integerMutableLiveData.getValue()) cnt = 0;
        }
    };

    public void getSummoner(String name) {
        mRiotApi.getUserInfo(name, BuildConfig.riot_api_key).enqueue(new Callback<SummonerDto>() {
            @Override
            public void onResponse(Call<SummonerDto> call, Response<SummonerDto> response) {
                if (response.isSuccessful() && response.body() != null) {
                    summonerDtoMutableLiveData.setValue(response.body());
                    getLeagueEntry();
                    getMatch(response.body().getPuuId());
                    getSpellJson();
                    getPerksJson();
                } else {
                    summonerDtoMutableLiveData.setValue(null);
                }
            }

            @Override
            public void onFailure(Call<SummonerDto> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void getLeagueEntry() {
        SummonerDto summonerDto = summonerDtoMutableLiveData.getValue();
        LeagueEntryDTO leagueEntryDTO = new LeagueEntryDTO();
        String id = summonerDto.getId();

        mRiotApi.getLeagueEntry(id, BuildConfig.riot_api_key).enqueue(
                new Callback<ArrayList<LeagueEntryDTO>>() {
                    @Override
                    public void onResponse(Call<ArrayList<LeagueEntryDTO>> call,
                                           Response<ArrayList<LeagueEntryDTO>> response) {
                        String path = "android.resource://" + R.class.getPackage().getName() + "/";
                        if (response.isSuccessful()) {
                            if (response.body().isEmpty()) {
                                path += getResourceId("UNRANK");
                                leagueEntryDTO.setTier("UNRANK");
                                leagueEntryDTO.setQueueType("");
                                leagueEntryDTO.setRank("");
                            } else {
                                path += getResourceId(response.body().get(0).getTier());
                                leagueEntryDTO.setTier(response.body().get(0).getTier());
                                leagueEntryDTO.setQueueType(response.body().get(0).getQueueType());
                                leagueEntryDTO.setRank(response.body().get(0).getRank());
                            }
                            summonerDto.setTierIconId(path);
                            summonerDto.setLeagueEntryDTO(leagueEntryDTO);
                            summonerDtoMutableLiveData.setValue(summonerDto);
                            getProfileIconImage();
                        }
                    }

                    @Override
                    public void onFailure(Call<ArrayList<LeagueEntryDTO>> call, Throwable t) {
                        t.printStackTrace();
                    }
                });
    }

    private int getResourceId(String resource) {
        switch (resource) {
            case "BRONZE":
                return R.drawable.emblem_bronze;
            case "IRON":
                return R.drawable.emblem_iron;
            case "SILVER":
                return R.drawable.emblem_silver;
            case "GOLD":
                return R.drawable.emblem_gold;
            case "PLATINUM":
                return R.drawable.emblem_platinum;
            case "DIAMOND":
                return R.drawable.emblem_diamond;
            case "MASTER":
                return R.drawable.emblem_master;
            case "GRANDMASTER":
                return R.drawable.emblem_grandmaster;
            case "CHALLENGER":
                return R.drawable.emblem_challenger;
            default:
                return R.drawable.emblem_unrank;
        }
    }

    private void getMatch(String puuId) {
        mRiotMatchApi.getMatchesId(puuId, BuildConfig.riot_api_key, 0, 19).
                enqueue(new Callback<ArrayList<String>>() {
                    @Override
                    public void onResponse(Call<ArrayList<String>> call,
                                           Response<ArrayList<String>> response) {
                        if (response.isSuccessful()) {
                            integerMutableLiveData.setValue(response.body().size());
                            for (String matchId : response.body()) {
                                getMatchInfo(matchId);
                            }
                        }
                    }

                    @Override
                    public void onFailure(Call<ArrayList<String>> call, Throwable t) {
                        t.printStackTrace();
                    }
                });
    }

    private String timeFormat(Long milliseconds) {
        String result = "";

        long hours = TimeUnit.MILLISECONDS.toHours(milliseconds);
        long minutes = TimeUnit.MILLISECONDS.toMinutes(milliseconds);
        long seconds = TimeUnit.MILLISECONDS.toSeconds(milliseconds);

        result = Long.toString(hours);
        result += Long.toString(minutes);
        result += Long.toString(seconds);

        return result;
    }

    public void getMatchInfo(String matchId) {
        mRiotMatchApi.getMatchInfo(matchId, BuildConfig.riot_api_key).enqueue(new Callback<MatchDto>() {
            @Override
            public void onResponse(Call<MatchDto> call, Response<MatchDto> response) {
                int pos = 0;
                if (response.isSuccessful()) {
                    MatchDto matchDto = response.body();
                    InfoDto infoDto = matchDto.getInfoDto();
                    ArrayList<ParticipantDto> arrayList = infoDto.getParticipantDto();
                    for (int i = 0; i < arrayList.size(); i++) {
                        if(arrayList.get(i).getSummonerName().equals(summonerDtoMutableLiveData.getValue().getName())) {
                            pos = i;
                            break;
                        }
                    }

                    ParticipantDto participantDto = infoDto.getParticipantDto().get(pos);
                    participantDto.setMatchId(matchId);

                    PerksDto perksDto = participantDto.getPerks();
                    PerksDto.PerkStyleDto mainPerkStyleDto = perksDto.getPerksStyleDto(0);
                    runeIds.add(mainPerkStyleDto.getSelections().get(0).getPerk());
                    PerksDto.PerkStyleDto subPerkStyleDto = perksDto.getPerksStyleDto(1);
                    runeIds.add(subPerkStyleDto.getStyle());
                    participantDto.setRuneIds(runeIds);

                    participantDtoMutableLiveData.setValue(participantDto);

                    itemIds.add(participantDto.getItem0());
                    itemIds.add(participantDto.getItem1());
                    itemIds.add(participantDto.getItem2());
                    itemIds.add(participantDto.getItem3());
                    itemIds.add(participantDto.getItem4());
                    itemIds.add(participantDto.getItem5());
                    itemIds.add(participantDto.getItem6());

                    spellIds.add(participantDto.getSummoner1Id());
                    spellIds.add(participantDto.getSummoner2Id());

                    apiListener.onUpdated();
                }
            }

            @Override
            public void onFailure(Call<MatchDto> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void getProfileIconImage() {
        SummonerDto summonerDto = summonerDtoMutableLiveData.getValue();
        String path = summonerDto.getProfileIconId() + ".png";

        mRiotProfileIconApi.getImage(path).enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                Bitmap bitmap = BitmapFactory.decodeStream(response.body().byteStream());
                summonerDto.setProfileIcon(bitmap);
                summonerDtoMutableLiveData.setValue(summonerDto);
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void getChampionImage(String name, String matchId) {
        String path = name + ".png";

        mRiotChampionImgApi.getImage(path).enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                Item item = new Item();
                item.setMatchId(matchId);

                Bitmap bitmap = BitmapFactory.decodeStream(response.body().byteStream());
                item.setItemBitmap(bitmap);

                championIconMutableLiveData.setValue(item);
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void getItemImage(int imgId, String matchId, int itemNum) {
        String path = imgId + ".png";

        mRiotItemImgApi.getImage(path).enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                Item item = new Item();
                item.setMatchId(matchId);
                item.setItemNum(itemNum);

                if (response.body() != null) {
                    Bitmap bitmap = BitmapFactory.decodeStream(response.body().byteStream());
                    item.setItemBitmap(bitmap);
                }
                else {
                    item.setItemBitmap(null);
                }
                itemIconMutableLiveData.setValue(item);
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void getSpellImage(int imgId, String matchId, int spellNum) {
        String spellName = spellNames.get(Integer.toString(imgId));
        String path = spellName + ".png";

        mRiotSpellImgApi.getImage(path).enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                Item item = new Item();
                item.setMatchId(matchId);
                item.setItemNum(spellNum);
                if (response.body() != null) {
                    Bitmap bitmap = BitmapFactory.decodeStream(response.body().byteStream());
                    item.setItemBitmap(bitmap);
                }
                else {
                    item.setItemBitmap(null);
                }
                spellIconMutableLiveData.setValue(item);
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void getSpellJson() {
        ArrayList<String> nameKeys = new ArrayList<>();

        mRiotSpellJSonApi.getSpellJson().enqueue(new Callback<Object>() {
            @Override
            public void onResponse(Call<Object> call, Response<Object> response) {
                String json = new Gson().toJson(response.body());
                try {
                    JSONObject jsonObject = new JSONObject(json);
                    jsonObject = jsonObject.getJSONObject("data");
                    Iterator iterator = jsonObject.keys();

                    while(iterator.hasNext()) {
                        String name = iterator.next().toString();
                        nameKeys.add(name);
                    }

                    for (int i = 0; i < nameKeys.size(); i++) {
                        JSONObject spell = jsonObject.getJSONObject(nameKeys.get(i));
                        String key = spell.getString("key");
                        spellNames.put(key, nameKeys.get(i));
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<Object> call, Throwable t) {

            }
        });
    }

    private void getPerksJson() {
        mRiotRunesJsonApi.getRuneJson().enqueue(new Callback<Object>() {
            @Override
            public void onResponse(Call<Object> call, Response<Object> response) {
                String json = new Gson().toJson(response.body());
                try {
                    JSONArray jsonArray = new JSONArray(json);
                    for (int i = 0; i < jsonArray.length(); i++) {
                        JSONObject mainRune = jsonArray.getJSONObject(i);
                        runeNames.put(mainRune.getInt("id"),
                                mainRune.getString("icon"));
                        JSONArray slots = mainRune.getJSONArray("slots");
                        for (int j = 0; j < slots.length(); j++) {
                            JSONObject runes = slots.getJSONObject(j);
                            JSONArray rune = runes.getJSONArray("runes");
                            for (int k = 0; k < rune.length(); k++) {
                                JSONObject jsonObject = rune.getJSONObject(k);
                                runeNames.put(jsonObject.getInt("id"),
                                        jsonObject.getString("icon"));
                            }
                        }
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(Call<Object> call, Throwable t) {

            }
        });
    }

    private void getRuneImage(int imgId, String matchId, int spellNum) {
        String path = runeNames.get(imgId);

        mRiotRuneImgApi.getImage(path).enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                Item item = new Item();
                item.setMatchId(matchId);
                item.setItemNum(spellNum);
                if (response.body() != null) {
                    Bitmap bitmap = BitmapFactory.decodeStream(response.body().byteStream());
                    item.setItemBitmap(bitmap);
                }
                else {
                    item.setItemBitmap(null);
                }
                runeIconMutableLiveData.setValue(item);
            }

            @Override
            public void onFailure(Call<ResponseBody> call, Throwable t) {
                t.printStackTrace();
            }
        });
    }
}

File: app/src/main/java/com/example/findmeinlol/RiotAPIRepository.kt
package com.example.findmeinlol

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.lifecycle.MutableLiveData
import com.example.findmeinlol.model.data.Match
import com.example.findmeinlol.model.data.Participant
import com.example.findmeinlol.model.data.Summoner
import com.example.findmeinlol.view.CallBack
import com.google.gson.Gson
import okhttp3.ResponseBody
import org.json.JSONArray
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

private interface ApiCallback {
    fun onUpdated()
}

data class Item(var matchId: String = "", var bitmap: Bitmap? = null, var number: Int = 0)

object RiotAPIRepository : ApiCallback {
    private const val RIOT_URL: String = "https://kr.api.riotgames.com/lol/";
    private const val RIOT_PROFILE_IMAGE_URL: String =
        "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/profileicon/";
    private const val RIOT_CHAMPION_IMAGE_URL: String =
        "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/champion/";
    private const val RIOT_ITEM_IMAGE_URL: String =
        "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/item/";
    private const val RIOT_MATCH_URL: String = "https://asia.api.riotgames.com/";
    private const val RIOT_JSON_URL: String = "https://ddragon.leagueoflegends.com/cdn/13.5.1/data/en_US/"
    private const val RIOT_SPELL_IMAGE_URL: String = "https://ddragon.leagueoflegends.com/cdn/13.3.1/img/spell/"
    private const val RIOT_RUNE_IMAGE_URL: String =  "https://ddragon.leagueoflegends.com/cdn/img/"

    var summonerLiveData: MutableLiveData<Summoner> = MutableLiveData()
    var participantLiveData: MutableLiveData<Participant> = MutableLiveData()
    var matchSizeLiveData: MutableLiveData<Int> = MutableLiveData()
    var championIconLiveData: MutableLiveData<Item> = MutableLiveData()
    var itemIconLiveData: MutableLiveData<Item> = MutableLiveData()
    var spellIconLiveData: MutableLiveData<Item> = MutableLiveData()
    var runeIconLiveData: MutableLiveData<Item> = MutableLiveData()

    private var itemIds: ArrayList<Int> = ArrayList()
    private var spellIds: ArrayList<Int> = ArrayList()
    private var runeIds: ArrayList<Int> = ArrayList()

    private var spellNames: HashMap<String, String> = hashMapOf()
    private var runeNames: HashMap<Int, String> = hashMapOf()

    private val riotAPI = Retrofit.Builder()
        .baseUrl(RIOT_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    private val riotProfileImageAPI = Retrofit.Builder()
        .baseUrl(RIOT_PROFILE_IMAGE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    private val riotMatchAPI = Retrofit.Builder()
        .baseUrl(RIOT_MATCH_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    private val riotChampionImageAPI = Retrofit.Builder()
        .baseUrl(RIOT_CHAMPION_IMAGE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    private val riotItemImageAPI = Retrofit.Builder()
        .baseUrl(RIOT_ITEM_IMAGE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    private val riotSpellImageApi = Retrofit.Builder()
        .baseUrl(RIOT_SPELL_IMAGE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    private val riotRuneImageApi = Retrofit.Builder()
        .baseUrl(RIOT_RUNE_IMAGE_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    private val riotJsonApi = Retrofit.Builder()
        .baseUrl(RIOT_JSON_URL)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
        .create(RiotApiKotlin::class.java)

    init {
        getJson()
    }

    private fun getJson() {
        riotJsonApi.getJson("summoner.json").enqueue(object: Callback<Any> {
            override fun onResponse(call: Call<Any>, response: Response<Any>) {
                val json = Gson().toJson(response.body())
                var jsonObject = JSONObject(json)
                jsonObject = jsonObject.getJSONObject("data")

                spellNames = jsonObject.keys().asSequence()
                    .associateBy(
                        { jsonObject.getJSONObject(it.toString()).getString("key")},
                        { it.toString() })
                        as HashMap<String, String>
            }

            override fun onFailure(call: Call<Any>, t: Throwable) {

            }
        })

        riotJsonApi.getJson("runesReforged.json").enqueue(object: Callback<Any> {
            override fun onResponse(call: Call<Any>, response: Response<Any>) {
                val json = Gson().toJson(response.body())
                val jsonArray = JSONArray(json)

                for (i in 0 until jsonArray.length()) {
                    val mainRune = jsonArray.getJSONObject(i)
                    runeNames[mainRune.getInt("id")] = mainRune.getString("icon")
                    val slots = mainRune.getJSONArray("slots")
                    for (j in 0 until slots.length()) {
                        val secondRune = slots.getJSONObject(j).getJSONArray("runes")
                        for (k in 0 until secondRune.length()) {
                            val rune = secondRune.getJSONObject(k)
                            runeNames[rune.getInt("id")] = rune.getString("icon")
                        }
                    }
                }
            }

            override fun onFailure(call: Call<Any>, t: Throwable) {
            }

        })
    }

    fun getSummoner(name: String) {
        riotAPI.getSummoner(name, BuildConfig.riot_api_key).enqueue(object : Callback<Summoner.SummonerDto> {
            override fun onResponse(
                call: Call<Summoner.SummonerDto>,
                response: Response<Summoner.SummonerDto>
            ) {
                var summoner = Summoner()
                summoner.summonerDto = response.body()
                summonerLiveData.value = summoner

                if (response.isSuccessful) {
                    getLeagueEntry()
                    getMatch()
                }
            }

            override fun onFailure(call: Call<Summoner.SummonerDto>, t: Throwable) {
            }
        })
    }

    private fun getLeagueEntry() {
        val id: String = summonerLiveData.value!!.summonerDto!!.id

        riotAPI.getLeagueEntry(id, BuildConfig.riot_api_key)
            .enqueue(object : Callback<ArrayList<Summoner.LeagueEntry>> {
                @RequiresApi(Build.VERSION_CODES.S)
                override fun onResponse(
                    call: Call<ArrayList<Summoner.LeagueEntry>>,
                    response: Response<ArrayList<Summoner.LeagueEntry>>
                ) {
                    var path: String =
                        "android.resource://" + R::class.java.packageName.toString() + "/";
                    var leagueEntry = Summoner.LeagueEntry()

                    if (response.isSuccessful) {
                        if (response.body()?.isEmpty() != true) {
                            leagueEntry = response.body()!![0]
                        }
                        path += getResourceId(leagueEntry.tier)
                        summonerLiveData.value!!.leagueEntry = leagueEntry
                        summonerLiveData.value!!.tierIconId = path

                        getIconImage()
                    }
                }

                override fun onFailure(call: Call<ArrayList<Summoner.LeagueEntry>>, t: Throwable) {
                    println(t.message)
                }
            })
    }

    private fun getIconImage() {
        var profileIconId: Int = summonerLiveData.value!!.summonerDto!!.profileIconId
        var path = """${profileIconId}.png"""

        riotProfileImageAPI.getImage(path).enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                if (response.isSuccessful) {
                    response.body()?.let {
                        summonerLiveData.value!!.profileIcon =
                            BitmapFactory.decodeStream(response.body()!!.byteStream())
                        summonerLiveData.value = summonerLiveData.value
                    }
                }
            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                println(t.message)
            }
        })
    }

    private fun getMatch() {
        val puuId: String = summonerLiveData.value!!.summonerDto!!.puuId

        riotMatchAPI.getMatchesId(puuId, BuildConfig.riot_api_key, 0, 19)
            .enqueue(object : Callback<ArrayList<String>> {
                override fun onResponse(
                    call: Call<ArrayList<String>>,
                    response: Response<ArrayList<String>>
                ) {
                    if (response.isSuccessful) {
                        response.body()?.let {
                            matchSizeLiveData.value = it.size
                            it.forEach { id -> getMatchDetail(id) }
                        }
                    }
                }
                override fun onFailure(call: Call<ArrayList<String>>, t: Throwable) {
                }
            })
    }

    private fun getMatchDetail(matchId: String) {
        riotMatchAPI.getMatchDetail(matchId, BuildConfig.riot_api_key)
            .enqueue(object : Callback<Match> {
                override fun onResponse(call: Call<Match>, response: Response<Match>) {
                    if (response.isSuccessful) {
                        response.body()?.let {
                            val participants: ArrayList<Participant.ParticipantDto> = it.info.participants

                            // 반복문 돌면서 조건에 맞는 첫번째 participant
                            val participantDto: Participant.ParticipantDto = participants.asSequence()
                                .filter { p -> p.summonerName == summonerLiveData.value!!.summonerDto!!.name }
                                .first()

                            var participant = Participant()
                            participant.gameDuration = getGameDuration(it.info.gameDuration)
                            participant.matchId = matchId
                            participant.participantDto = participantDto
                            participantLiveData.value = participant

                            setIds()
                            onUpdated()
                        }
                    }
                }

                override fun onFailure(call: Call<Match>, t: Throwable) {
                }
            })
    }

    private fun getChampionIconImage(matchId: String, championName: String) {
        val path = """${championName}.png"""
        var item = Item()

        riotChampionImageAPI.getImage(path).enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                if (response.isSuccessful) {
                    response.body()?.let {
                        item.matchId = matchId
                        item.bitmap = BitmapFactory.decodeStream(it.byteStream())
                        championIconLiveData.value = item
                    }
                }
            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
            }
        })
    }

    private fun getItemImage(matchId: String, itemId: Int, pos: Int) {
        val path = """${itemId}.png"""

        riotItemImageAPI.getImage(path).enqueue(object : Callback<ResponseBody> {
            override fun onResponse(
                call: Call<ResponseBody>,
                response: Response<ResponseBody>
            ) {
                var item = Item()
                item.matchId = matchId
                item.number = pos

                if (response.isSuccessful) {
                    item.bitmap = BitmapFactory.decodeStream(response.body()!!.byteStream())
                } else {
                    item.bitmap = null
                }
                itemIconLiveData.value = item
            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
            }

        })
    }

    private fun getSpellImage(matchId: String, spellId: Int, pos: Int) {
        val path: String = spellNames[spellId.toString()] + ".png"

        riotSpellImageApi.getImage(path).enqueue(object: Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                var item = Item()
                item.matchId = matchId
                item.number = pos

                if (response.isSuccessful) {
                    item.bitmap = BitmapFactory.decodeStream(response.body()!!.byteStream())
                }
                else {
                    item.bitmap = null
                }

                spellIconLiveData.value = item
            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {

            }
        })
    }

    private fun getRuneImage(matchId: String, runeId: Int, pos: Int) {
        val path: String = runeNames[runeId].toString()

        riotRuneImageApi.getImage(path).enqueue(object: Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                var item = Item()
                item.matchId = matchId
                item.number = pos

                if (response.isSuccessful) {
                    item.bitmap = BitmapFactory.decodeStream(response.body()!!.byteStream())
                }
                else {
                    item.bitmap = null
                }

                runeIconLiveData.value = item
            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {

            }
        })
    }

    private fun setIds() {
        itemIds.add(participantLiveData.value!!.participantDto!!.item0)
        itemIds.add(participantLiveData.value!!.participantDto!!.item1)
        itemIds.add(participantLiveData.value!!.participantDto!!.item2)
        itemIds.add(participantLiveData.value!!.participantDto!!.item3)
        itemIds.add(participantLiveData.value!!.participantDto!!.item4)
        itemIds.add(participantLiveData.value!!.participantDto!!.item5)
        itemIds.add(participantLiveData.value!!.participantDto!!.item6)

        spellIds.add(participantLiveData.value!!.participantDto!!.summoner1Id)
        spellIds.add(participantLiveData.value!!.participantDto!!.summoner2Id)

        runeIds.add(participantLiveData.value!!.participantDto!!.perks.styles[0].selections[0].perk)
        runeIds.add(participantLiveData.value!!.participantDto!!.perks.styles[1].style)
    }

    private fun getResourceId(resource: String): Int {
        return when (resource) {
            "BRONZE" -> R.drawable.emblem_bronze
            "IRON" -> R.drawable.emblem_iron
            "SLIVER" -> R.drawable.emblem_silver
            "GOLD" -> R.drawable.emblem_gold
            "PLATINUM" -> R.drawable.emblem_platinum
            "DIAMOND" -> R.drawable.emblem_diamond
            "MASTER" -> R.drawable.emblem_master
            "GRANDMASTER" -> R.drawable.emblem_grandmaster
            "CHALLENGER" -> R.drawable.emblem_challenger
            else -> {
                R.drawable.emblem_unrank
            }
        }
    }

    private fun getGameDuration(gameDuration: Long): String {
        val minute = TimeUnit.SECONDS.toMinutes(gameDuration)
        - TimeUnit.SECONDS.toHours(gameDuration) * 60
        val second = gameDuration - minute * 60

        return """${minute}:${second}"""
    }

    override fun onUpdated() {
        val participant: Participant = participantLiveData.value!!

        getChampionIconImage(participant.matchId, participant.participantDto!!.championName)
        for (i in 0..6) {
            getItemImage(participant.matchId, itemIds[i], i)
        }
        for (i in 0..1) {
            getSpellImage(participant.matchId, spellIds[i], i)
        }
        for (i in 0..1) {
            getRuneImage(participant.matchId, runeIds[i], i)
        }

        itemIds.clear()
        runeIds.clear()
        spellIds.clear()
    }
}